File: java/src/test/java/com/google/crypto/tink/hybrid/HybridEncryptCatalogueTest.java
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.hybrid;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

import com.google.crypto.tink.HybridEncrypt;
import com.google.crypto.tink.KeyManager;
import com.google.crypto.tink.proto.KeyTypeEntry;
import com.google.crypto.tink.proto.RegistryConfig;
import java.security.GeneralSecurityException;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/** Tests for HybridEncryptCatalogue. */
@RunWith(JUnit4.class)
public class HybridEncryptCatalogueTest {

  @Test
  public void testBasic() throws Exception {
    HybridEncryptCatalogue catalogue = new HybridEncryptCatalogue();

    // Check a single key type for encryption, incl. case-insensitve primitive name.
    String keyType = "type.googleapis.com/google.crypto.tink.EciesAeadHkdfPublicKey";
    {
      KeyManager<HybridEncrypt> manager = catalogue.getKeyManager(keyType, "HybridEncrypt", 0);
      assertThat(manager.doesSupport(keyType)).isTrue();
    }
    {
      KeyManager<HybridEncrypt> manager = catalogue.getKeyManager(keyType, "HybRIdEncRYPt", 0);
      assertThat(manager.doesSupport(keyType)).isTrue();
    }
    {
      KeyManager<HybridEncrypt> manager = catalogue.getKeyManager(keyType, "HYBRIdeNCRYPT", 0);
      assertThat(manager.doesSupport(keyType)).isTrue();
    }

    // Check all entries from the current HybridConfig.
    RegistryConfig config = HybridConfig.TINK_1_0_0;
    int count = 0;
    for (KeyTypeEntry entry : config.getEntryList()) {
      if ("HybridEncrypt".equals(entry.getPrimitiveName())) {
        count = count + 1;
        KeyManager<HybridEncrypt> manager = catalogue.getKeyManager(
            entry.getTypeUrl(), "hybridencrypt", entry.getKeyManagerVersion());
        assertThat(manager.doesSupport(entry.getTypeUrl())).isTrue();
      }
    }
    assertEquals(1, count);
  }

  @Test
  public void testErrors() throws Exception {
    HybridEncryptCatalogue catalogue = new HybridEncryptCatalogue();
    String keyType = "type.googleapis.com/google.crypto.tink.EciesAeadHkdfPublicKey";

    // Wrong primitive name.
    try {
      catalogue.getKeyManager(keyType, "aead", 0);
      fail("Expected GeneralSecurityException");
    } catch (GeneralSecurityException e) {
      assertThat(e.toString()).contains("No support for primitive");
    }

    // Wrong key manager version.
    try {
      catalogue.getKeyManager(keyType, "hybridencrypt", 1);
      fail("Expected GeneralSecurityException");
    } catch (GeneralSecurityException e) {
      assertThat(e.toString()).contains("No key manager");
      assertThat(e.toString()).contains("version at least");
    }

    // Wrong key type.
    try {
      catalogue.getKeyManager("some.unknown.key.type", "hybridencrypt", 0);
      fail("Expected GeneralSecurityException");
    } catch (GeneralSecurityException e) {
      assertThat(e.toString()).contains("No support for primitive");
      assertThat(e.toString()).contains("with key type");
    }
  }
}

File: javascript/hybrid/hybrid_encrypt_catalogue_test.js
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

goog.module('tink.hybrid.HybridEncryptCatalogueTest');
goog.setTestOnly('tink.hybrid.HybridEncryptCatalogueTest');

const EciesAeadHkdfPublicKeyManager = goog.require('tink.hybrid.EciesAeadHkdfPublicKeyManager');
const HybridEncryptCatalogue = goog.require('tink.hybrid.HybridEncryptCatalogue');

const testSuite = goog.require('goog.testing.testSuite');

const SUPPORTED_PRIMITIVE_NAME = 'HybridEncrypt';

testSuite({
  testGetKeyManager_wrongPrimitive() {
    const anotherPrimitiveName = 'SOME_UNSUPPORTED_PRIMITIVE';

    const catalogue = new HybridEncryptCatalogue();
    try {
      catalogue.getKeyManager(
          EciesAeadHkdfPublicKeyManager.KEY_TYPE, anotherPrimitiveName,
          /* minVersion = */ 0);
      fail('An exception should be thrown.');
    } catch (e) {
      assertEquals(
          ExceptionText.wrongPrimitive(anotherPrimitiveName), e.toString());
    }
  },

  testGetKeyManager_versionOutOfBounds() {
    const manager = new EciesAeadHkdfPublicKeyManager();
    const version = manager.getVersion() + 1;

    const catalogue = new HybridEncryptCatalogue();
    try {
      catalogue.getKeyManager(
          manager.getKeyType(), SUPPORTED_PRIMITIVE_NAME, version);
      fail('An exception should be thrown.');
    } catch (e) {
      assertEquals(
          ExceptionText.versionOutOfBounds(manager.getVersion()), e.toString());
    }
  },

  testGetKeyManager_unknownKeyType() {
    const keyType = 'UNKNOWN_KEY_TYPE';
    const version = 0;

    const catalogue = new HybridEncryptCatalogue();
    try {
      catalogue.getKeyManager(keyType, SUPPORTED_PRIMITIVE_NAME, version);
      fail('An exception should be thrown.');
    } catch (e) {
      assertEquals(ExceptionText.unknownKeyType(keyType), e.toString());
    }
  },

  testGetKeyManager_eciesAeadHkdfPublicKeyManager() {
    const catalogue = new HybridEncryptCatalogue();
    const version = 0;

    const manager = catalogue.getKeyManager(
        EciesAeadHkdfPublicKeyManager.KEY_TYPE, SUPPORTED_PRIMITIVE_NAME,
        version);

    assertObjectEquals(new EciesAeadHkdfPublicKeyManager(), manager);
  },

  testGetKeyManager_caseInsensitivePrimitiveName() {
    const catalogue = new HybridEncryptCatalogue();
    const version = 0;
    const keyType = EciesAeadHkdfPublicKeyManager.KEY_TYPE;

    catalogue.getKeyManager(keyType, 'hybridEncrypt', version);
    catalogue.getKeyManager(keyType, 'HybridEncrypt', version);
    catalogue.getKeyManager(keyType, 'hybridencrypt', version);
    catalogue.getKeyManager(keyType, 'HYBRIDENCRYPT', version);
  },
});

// Helper classes and functions
class ExceptionText {
  /**
   * @param {string} requested
   * @return {string}
   */
  static wrongPrimitive(requested) {
    return 'CustomError: Requested ' + requested + ' primitive, but this ' +
        'catalogue provides key managers for ' + SUPPORTED_PRIMITIVE_NAME +
        ' primitives.';
  }

  /**
   * @param {number} availableVersion
   * @return {string}
   */
  static versionOutOfBounds(availableVersion) {
    return 'CustomError: Requested manager with higher version ' +
        'than is available. The available manager has version ' +
        availableVersion + '.';
  }

  /**
   * @param {string} keyType
   * @return {string}
   */
  static unknownKeyType(keyType) {
    return 'CustomError: There is no key manager for key type ' + keyType +
        ' available.';
  }
}
